contrib/babelfishpg_tsql/src/catalog.c

```diff
@@ -2799,6 +2828,614 @@ rename_procfunc_update_bbf_catalog(RenameStmt *stmt)
 	table_close(bbf_func_ext_rel, RowExclusiveLock);
 }
 
+/*
+ * Add an entry to catalog BABELFISH_SCHEMA_PERMISSIONS.
+ */
+void
+add_entry_to_bbf_schema_perms(const char *schema_name,
+				const char *object_name,
+				int permission,
+				const char *grantee,
+				const char *object_type)
+{
+	Relation	bbf_schema_rel;
+	TupleDesc	bbf_schema_dsc;
+	HeapTuple	tuple_bbf_schema;
+	Datum		new_record_bbf_schema[BBF_SCHEMA_PERMS_NUM_OF_COLS];
+	bool		new_record_nulls_bbf_schema[BBF_SCHEMA_PERMS_NUM_OF_COLS];
+	int16	dbid = get_cur_db_id();
+
+	/* Immediately return, if grantee is NULL or PUBLIC. */
+	if ((grantee == NULL) || (strncmp(grantee, "public", NAMEDATALEN) == 0))
+		return;
+
+	/* Fetch the relation */
+	bbf_schema_rel = table_open(get_bbf_schema_perms_oid(),
+									RowExclusiveLock);
+	bbf_schema_dsc = RelationGetDescr(bbf_schema_rel);
+
+	/* Build a tuple to insert */
+	MemSet(new_record_bbf_schema, 0, sizeof(new_record_bbf_schema));
+	MemSet(new_record_nulls_bbf_schema, false, sizeof(new_record_nulls_bbf_schema));
+
+	new_record_bbf_schema[BBF_SCHEMA_PERMS_DBID] = Int16GetDatum(dbid);
+	new_record_bbf_schema[BBF_SCHEMA_PERMS_SCHEMA_NAME] = CStringGetDatum(pstrdup(schema_name));
+	new_record_bbf_schema[BBF_SCHEMA_PERMS_OBJECT_NAME] = CStringGetDatum(pstrdup(object_name));
+	new_record_bbf_schema[BBF_SCHEMA_PERMS_PERMISSION] = Int32GetDatum(permission);
+	new_record_bbf_schema[BBF_SCHEMA_PERMS_GRANTEE] = CStringGetDatum(pstrdup(grantee));
+	new_record_bbf_schema[BBF_SCHEMA_PERMS_OBJECT_TYPE] = CStringGetDatum(pstrdup(object_type));
+
+	tuple_bbf_schema = heap_form_tuple(bbf_schema_dsc,
+									new_record_bbf_schema,
+									new_record_nulls_bbf_schema);
+
+	/* Insert new record in the bbf_authid_user_ext table */
+	CatalogTupleInsert(bbf_schema_rel, tuple_bbf_schema);
+
+	/* Close bbf_authid_user_ext, but keep lock till commit */
+	table_close(bbf_schema_rel, RowExclusiveLock);
+
+	/* Advance cmd counter to make the insert visible */
+	CommandCounterIncrement();
```

Comment:
Over-use of commandCounterIncrement, why would we require the tuple to be visible just after? Wont it be just a single grant statement in that query? Also in code flow we will CommandCounterIncrement more than once, can you validate?

Reply:
In case of grant statement with multiple grantees, this will be executed that number of times.